src/ripple/app/paths/impl/BookStep.cpp

```diff
@@ -237,60 +259,104 @@ void limitStepOut (Quality const& ofrQ,
-        ? rate (book.out.account)
+    std::uint32_t const trOut = ownerPaysTransferFee_
+        ? rate (book_.out.account)
         : QUALITY_ONE;
 
-    typename FlowOfferStream<TAmtIn, TAmtOut>::StepCounter counter (limit, j);
-    FlowOfferStream<TAmtIn, TAmtOut> offers (
-        sb, afView, book, sb.parentCloseTime (), counter, j);
+    typename FlowOfferStream<TIn, TOut>::StepCounter
+        counter (maxOffersToConsume_, j_);
+
+    FlowOfferStream<TIn, TOut> offers (
+        sb, afView, book_, sb.parentCloseTime (), counter, j_);
 
+    bool offerAttempted = false;
     boost::optional<Quality> ofrQ;
     while (offers.step ())
     {
         auto& offer = offers.tip ();
+
+        // Note that offer.quality() returns a (non-optional) Quality.  So
+        // ofrQ is always safe to use below this point in the loop.
         if (!ofrQ)
             ofrQ = offer.quality ();
         else if (*ofrQ != offer.quality ())
             break;
 
+        // Special rule for offer crossing.  If:
+        //   a. We're offer crossing using default path (no autobridging), and
+        //   b. The offer's quality is at least as good as our quality, and
+        //   c. We're about to cross one of our own offers, then
+        //   c. Delete the old offer.
+        // See explanation in the testSelfCrossOffer1() unit test.
+        if (limitSelfCrossingQuality_ &&
+            *ofrQ >= *limitSelfCrossingQuality_ &&
+            strandSrc_ == offer.owner() &&
+            strandDst_ == offer.owner())
+        {
+            // Remove this offer even if no crossing occurs.
+            offers.permRmOffer (offer.key());
+
+            // If no offers have been attempted yet it's okay to move to
+            // a different quality.
+            if (!offerAttempted)
+                ofrQ = boost::none;
+
+            // Note: continue deletes the offer when offers.step() is called.
+            continue;
+        }
+
+        // For compatibility with V1 offer crossing, in offer crossing
+        // (only) we don't pay the transfer fee if alice is paying alice.
+        boost::optional<AccountID> const srcAcct = [this] ()
+        {
+            if (this->prevStep_)
+                return this->prevStep_->directStepSrcAcct();
+            return boost::optional<AccountID>();
+        }();
+        auto const ofrInRate =
+            noXferFeeIfSourceIsDest_ &&               // If offer crossing
+            srcAcct &&                                // && prevStep is DirectI
+            offer.owner() == *srcAcct                 // && src is offer owner
+            ? QUALITY_ONE : trIn;                     // then rate = QUALITY_ONE
+
+        auto const ofrOutRate =
+            noXferFeeIfSourceIsDest_ &&               // If offer crossing
+            prevStep_ && prevStep_->bookStepBook() && // && prevStep is BookStep
+            offer.owner() == strandDst_               // && dest is offer owner
```

Comment:
Checking `offer.owner() == strandDst_` makes sense if there is only one step after this one. That's true in offer x-ing, but it's not clear from the context that this is offer x-ing only logic (`noXferFeeIfSourceIsDest_ will only be set of offer x-ing, but somebody trying to understand this code and reading this line is going to think it's wrong). I think this is an argument for a separate offer x-ing book step, in which case this logic for `ofrOutRate` would go in a virtual function that _clearly_ for offer x-ing only.